File: 2352-equal-row-and-column-pairs/2352-equal-row-and-column-pairs.js
/**
 * @param {number[][]} grid
 * @return {number}
 */
var equalPairs = function(grid) {
    const n = grid.length
    const cnts = {}
    for (const row of grid) {
        const key = "|" + row.join("|")
        cnts[key] = cnts[key] + 1 || 1
    }
    let res = 0
    for (let c=0; c < n; c++) {
        let key = ""
        for (let r=0; r < n; r++) {
            key += "|" + grid[r][c]
        }
        res += cnts[key] || 0
    }
    return res
};

equalPairs = function(grid) {
    class Trie {
        constructor() {
            this.children = {}
            this.count = 0
            this.end = false
        }
        insert(arr) {
            let cur = this
            for (const n of arr) {
                if (!(n in cur.children)) {
                    cur.children[n] = new Trie()
                }
                cur.children[n].count++
                cur = cur.children[n]
            }
            cur.end = true
        }
        find(arr) {
            let cur = this
            for (const n of arr) {
                if (n in cur.children) {
                    cur = cur.children[n]
                } else {
                    return 0
                }
            }
            return cur.count
        }
    }
    
    const n = grid.length
    const trie = new Trie()
    for (const row of grid) {
        trie.insert(row)
    }
    let res = 0
    for (let c=0; c < n; c++) {
        const arr = []
        for (let r=0; r < n; r++) {
            arr.push(grid[r][c])
        }
        res += trie.find(arr)
    }
    return res
}

File: 2352-equal-row-and-column-pairs/2352-equal-row-and-column-pairs.py
class Trie:
    def __init__(self):
        self.children = {}
        self.count = 0
    
    def insert(self, arr: List[int]) -> None:
        cur = self
        for n in arr:
            if str(n) not in cur.children:
                cur.children[str(n)] = Trie()
            cur.children[str(n)].count += 1
            cur = cur.children[str(n)]
    
    def find(self, arr: List[int]) -> int:
        cur = self
        for n in arr:
            if str(n) in cur.children:
                cur = cur.children[str(n)]
            else:
                return 0
        return cur.count

class Solution:
    def equalPairs(self, grid: List[List[int]]) -> int:
        n, trie, res = len(grid), Trie(), 0
        for row in grid:
            trie.insert(row)
        for c in range(0, n):
            arr = []
            for r in range(0, n):
                arr.append(grid[r][c])
            res += trie.find(arr)
        return res
            
            
        n, cnt, res = len(grid), {}, 0
        for row in grid:
            key = "|" + "|".join([str(c) for c in row])
            cnt[key] = cnt.get(key, 0) + 1
        for c in range(0, n):
            key = ""
            for r in range(0, n):
                key += "|" + str(grid[r][c])
            res += cnt.get(key, 0)
        return res
        